### DISCUSSION TITLE
drivers:serial Unable to achieve continuous reception on STM32 with async API

### DISCUSSION THREAD

**Discussion by msmttchr** [zephyrproject-rtos/zephyr#79831] [General]
We are having trouble using some of the lower-end MCUs (STM32F0, G0 and C0) for some of our end-devices.
The MCUs in these devices must support a UART interface at sustained 500Kbps with no flow control. 
We've noticed data loss when using Zephyr v3.7.0 UART driver with UART_ASYNC_API.
Using this configuration we have seen data loss when the DMA interrupts and attempts to restart the DMA with a new buffer.
We believe the cause of the data loss is because the current driver implementation doesn’t process data quickly enough.

Normally, STM32 supports circular dma buffer management with ability to have half full irq and full irq, so that software can read previous half while current half is being filled. In this case, no dma buffer setup is needed during reception, but I believe Zephyr drivers model does not support circular buffer.

Could you please suggest how to handle this use case?

**Comment by erwango** [comment]
@djiatsaf-st ^^

**Comment by djiatsaf-st** [comment]
Hello @msmttchr,
I have read the problem encountered, and i think it is possible to have data loss in this situation with the current version of the driver. 
A workaround would be to use the circular buffer as suggested.
Also Zephyr supports the circular buffer as "ring buffer" i think.
Now i am working on it to add/enable the circular mode on stm32 driver with DMA which will allow to use this mode on STM32F0, G0, C0 series in particular.
I hope to be able to make this possible very soon

**Comment by djiatsaf-st** [comment]
Update 

PR available https://github.com/zephyrproject-rtos/zephyr/pull/80921
